File: plugin/src/test/java/com/stratio/cassandra/lucene/schema/analysis/tokenizer/TokenizerBuilderTest.java
package com.stratio.cassandra.lucene.schema.analysis.tokenizer;

import com.google.common.collect.Sets;
import com.stratio.cassandra.lucene.common.JsonSerializer;
import org.apache.lucene.analysis.Tokenizer;
import org.apache.lucene.analysis.core.*;
import org.apache.lucene.analysis.ngram.EdgeNGramTokenizer;
import org.apache.lucene.analysis.ngram.NGramTokenizer;
import org.apache.lucene.analysis.path.PathHierarchyTokenizer;
import org.apache.lucene.analysis.path.ReversePathHierarchyTokenizer;
import org.apache.lucene.analysis.pattern.PatternTokenizer;
import org.apache.lucene.analysis.standard.ClassicTokenizer;
import org.apache.lucene.analysis.standard.StandardTokenizer;
import org.apache.lucene.analysis.standard.UAX29URLEmailTokenizer;
import org.apache.lucene.analysis.th.ThaiTokenizer;
import org.apache.lucene.analysis.wikipedia.WikipediaTokenizer;
import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.*;

/**
 * @author Eduardo Alonso {@literal <eduardoalonso@stratio.com>}
 */
public class TokenizerBuilderTest {

    private <T> T assertBuilderAndTokenizer(String json, Class expectedBuilderClass, Class expectedTokenizerClass) {
        try {
            TokenizerBuilder abstractBuilder = JsonSerializer.fromString(json, TokenizerBuilder.class);
            assertEquals("Expected " + expectedBuilderClass.getName() + " class",
                         expectedBuilderClass,
                         abstractBuilder.getClass());
            Tokenizer tokenizer = abstractBuilder.buildTokenizer();
            assertEquals("Expected " + expectedTokenizerClass.getName() + " class",
                         expectedTokenizerClass,
                         tokenizer.getClass());
            return (T) abstractBuilder;
        } catch (Exception e) {
            fail(e.getLocalizedMessage());
            return null;
        }
    }

    private void assertJsonParseFail(String json) throws IOException {
        JsonSerializer.fromString(json, TokenizerBuilder.class);
    }

    private void assertJsonParseFail(String json, String message) {
        try {
            JsonSerializer.fromString(json, TokenizerBuilder.class);
        } catch (IOException e) {
            assertEquals("Expected IOException with message: " +
                         message +
                         " but received: " +
                         e.getMessage() +
                         " localMess: " +
                         e.getLocalizedMessage(), message, e.getMessage());
        }
        assertFalse("Parsing: " + json + " must generate an IOException with message: " + message + " but does not.",
                    true);
    }

    private void assertExactValue(String paramName, Object expected, Object received) {
        assertEquals("Expected " +
                     paramName +
                     " equals to " +
                     expected.toString() +
                     " but received: " +
                     received.toString(), expected, received);
    }

    @Test
    public void testClassicTokenizerValidJSON() {
        String json = "{type: \"classic\", max_token_length: 250}";
        ClassicTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                    ClassicTokenizerBuilder.class,
                                                                    ClassicTokenizer.class);
        assertExactValue("ClassicTokenizerBuilder.maxTokenLength", 250, builder.maxTokenLength);
    }

    @Test
    public void testClassicTokenizerDefaultValues() {
        ClassicTokenizerBuilder builder = assertBuilderAndTokenizer("{type: \"classic\"}",
                                                                    ClassicTokenizerBuilder.class,
                                                                    ClassicTokenizer.class);
        assertExactValue("ClassicTokenizerBuilder.maxTokenLength",
                         ClassicTokenizerBuilder.DEFAULT_MAX_TOKEN_LENGTH,
                         builder.maxTokenLength);
    }

    @Test(expected = IOException.class)
    public void testClassicTokenizerInvalidParam() throws IOException {
        assertJsonParseFail("{type: \"classic\", max_toen_length: 250}");
    }

    @Test
    public void testKeywordTokenizerValidJSON() {
        String json = "{type: \"keyword\", buffer_size: 256}";
        KeywordTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                    KeywordTokenizerBuilder.class,
                                                                    KeywordTokenizer.class);
        assertExactValue("KeywordTokenizer.bufferSize", 256, builder.bufferSize);
    }

    @Test
    public void testKeywordTokenizerDefaultValues() {
        KeywordTokenizerBuilder builder = assertBuilderAndTokenizer("{type: \"keyword\"}",
                                                                    KeywordTokenizerBuilder.class,
                                                                    KeywordTokenizer.class);
        assertExactValue("ClassicTokenizerBuilder.maxTokenLength",
                         KeywordTokenizerBuilder.DEFAULT_BUFFER_SIZE,
                         builder.bufferSize);
    }

    @Test(expected = IOException.class)
    public void testKeywordTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"keyword\", bufer_size: 256}");
    }

    @Test
    public void testLetterTokenizerValidJSON() {
        assertBuilderAndTokenizer("{type: \"letter\"}", LetterTokenizerBuilder.class, LetterTokenizer.class);
    }

    @Test
    public void testLowerCaseTokenizerValidJSON() {
        assertBuilderAndTokenizer("{type: \"lower_case\"}", LowerCaseTokenizerBuilder.class, LowerCaseTokenizer.class);
    }

    @Test
    public void testThaiTokenizerValidJSON() {
        assertBuilderAndTokenizer("{type: \"thai\"}", ThaiTokenizerBuilder.class, ThaiTokenizer.class);
    }

    @Test
    public void testNGramTokenizerValidJSON() {
        String json = "{type: \"ngram\", min_gram: 1, max_gram: 2}";
        NGramTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                  NGramTokenizerBuilder.class,
                                                                  NGramTokenizer.class);
        assertExactValue("NGramTokenizerBuilder.min_gram", NGramTokenizerBuilder.DEFAULT_MIN_GRAM, builder.minGram);
        assertExactValue("NGramTokenizerBuilder.max_gram", NGramTokenizerBuilder.DEFAULT_MAX_GRAM, builder.maxGram);
    }

    @Test
    public void testNGramTokenizerDefaultValues() {
        String json = "{type: \"ngram\"}";
        NGramTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                  NGramTokenizerBuilder.class,
                                                                  NGramTokenizer.class);
        assertExactValue("NGramTokenizerBuilder.min_gram", NGramTokenizerBuilder.DEFAULT_MIN_GRAM, builder.minGram);
        assertExactValue("NGramTokenizerBuilder.max_gram", NGramTokenizerBuilder.DEFAULT_MAX_GRAM, builder.maxGram);
    }

    @Test(expected = IOException.class)
    public void testNGramTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"ngram\", min_am: 1, max_gram: 1}");
    }

    @Test
    public void testEdgeNGramTokenizerValidJSON() {
        String json = "{type: \"edge_ngram\", min_gram: 1, max_gram: 1}";
        EdgeNGramTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                      EdgeNGramTokenizerBuilder.class,
                                                                      EdgeNGramTokenizer.class);
        assertExactValue("EdgeNGramTokenizerBuilder.min_gram",
                         EdgeNGramTokenizer.DEFAULT_MIN_GRAM_SIZE,
                         builder.minGram);
        assertExactValue("EdgeNGramTokenizerBuilder.max_gram",
                         EdgeNGramTokenizer.DEFAULT_MAX_GRAM_SIZE,
                         builder.maxGram);
    }

    @Test
    public void testEdgeNGramTokenizerDefaultValues() {
        String json = "{type: \"edge_ngram\"}";
        EdgeNGramTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                      EdgeNGramTokenizerBuilder.class,
                                                                      EdgeNGramTokenizer.class);
        assertExactValue("EdgeNGramTokenizerBuilder.min_gram",
                         EdgeNGramTokenizer.DEFAULT_MIN_GRAM_SIZE,
                         builder.minGram);
        assertExactValue("EdgeNGramTokenizerBuilder.max_gram",
                         EdgeNGramTokenizer.DEFAULT_MAX_GRAM_SIZE,
                         builder.maxGram);
    }

    @Test(expected = IOException.class)
    public void testEdgeNGramTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"edge_ngram\", min_am: 1, max_gram: 1}");
    }

    @Test
    public void testPathHierarchyTokenizerValidJSON() {
        String json = "{type: \"path_hierarchy\", buffer_size: 246, delimiter: \"$\", replacement: \"%\", skip: 3}";
        PathHierarchyTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                          PathHierarchyTokenizerBuilder.class,
                                                                          PathHierarchyTokenizer.class);
        assertExactValue("PathHierarchyTokenizerBuilder.buffer_size", 246, builder.bufferSize);
        assertExactValue("PathHierarchyTokenizerBuilder.delimiter", '$', builder.delimiter);
        assertExactValue("PathHierarchyTokenizerBuilder.replacement", '%', builder.replacement);
        assertExactValue("PathHierarchyTokenizerBuilder.skip", 3, builder.skip);
    }

    @Test
    public void testPathHierarchyTokenizerDefaultValues() {
        String json = "{type: \"path_hierarchy\"}";
        PathHierarchyTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                          PathHierarchyTokenizerBuilder.class,
                                                                          PathHierarchyTokenizer.class);
        assertExactValue("PathHierarchyTokenizerBuilder.buffer_size",
                         PathHierarchyTokenizerBuilder.DEFAULT_BUFFER_SIZE,
                         builder.bufferSize);
        assertExactValue("PathHierarchyTokenizerBuilder.delimiter",
                         PathHierarchyTokenizerBuilder.DEFAULT_DELIMITER,
                         builder.delimiter);
        assertExactValue("PathHierarchyTokenizerBuilder.replacement",
                         PathHierarchyTokenizerBuilder.DEFAULT_REPLACEMENT,
                         builder.replacement);
        assertExactValue("PathHierarchyTokenizerBuilder.skip",
                         PathHierarchyTokenizerBuilder.DEFAULT_SKIP,
                         builder.skip);
    }

    @Test(expected = IOException.class)
    public void testPathHierarchyTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"path_hierarchy\", buffer_size: 246, delimter: \"$\", replacement: \"%\", skip: 3}");
    }

    @Test
    public void testPatternTokenizerValidJSON() {
        String json = "{type: \"pattern\", pattern: \"[a-z]\", flags: 35, group: 0}";
        PatternTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                    PatternTokenizerBuilder.class,
                                                                    PatternTokenizer.class);
        assertExactValue("PathHierarchyTokenizerBuilder.pattern", "[a-z]", builder.pattern);
        assertExactValue("PathHierarchyTokenizerBuilder.flags", 35, builder.flags);
        assertExactValue("PathHierarchyTokenizerBuilder.group", 0, builder.group);
    }

    @Test
    public void testPatternTokenizerDefaultValues() {
        String json = "{type: \"pattern\"}";
        PatternTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                    PatternTokenizerBuilder.class,
                                                                    PatternTokenizer.class);
        assertExactValue("PathHierarchyTokenizerBuilder.pattern",
                         PatternTokenizerBuilder.DEFAULT_PATTERN,
                         builder.pattern);
        assertExactValue("PathHierarchyTokenizerBuilder.group", PatternTokenizerBuilder.DEFAULT_GROUP, builder.group);
        assertExactValue("PathHierarchyTokenizerBuilder.group", PatternTokenizerBuilder.DEFAULT_FLAGS, builder.flags);
    }

    @Test(expected = IOException.class)
    public void testPatternTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"pattern\", paern: \"[a-z]\", flags: 35, group: 0}");
    }

    @Test
    public void testReversePathHierarchyTokenizerValidJSON() {
        String
                json
                = "{type: \"reverse_path_hierarchy\", buffer_size: 246, delimiter: \"/\", replacement: \"%\", skip: 3}";
        ReversePathHierarchyTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                                 ReversePathHierarchyTokenizerBuilder.class,
                                                                                 ReversePathHierarchyTokenizer.class);
        assertExactValue("ReversePathHierarchyTokenizerBuilder.buffer_size", 246, builder.bufferSize);
        assertExactValue("ReversePathHierarchyTokenizerBuilder.delimiter", '/', builder.delimiter);
        assertExactValue("ReversePathHierarchyTokenizerBuilder.replacement", '%', builder.replacement);
        assertExactValue("ReversePathHierarchyTokenizerBuilder.skip", 3, builder.skip);
    }

    @Test
    public void testReversePathHierarchyTokenizerDefaultValues() {
        String json = "{type: \"reverse_path_hierarchy\"}";
        ReversePathHierarchyTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                                 ReversePathHierarchyTokenizerBuilder.class,
                                                                                 ReversePathHierarchyTokenizer.class);
        assertExactValue("PathHierarchyTokenizerBuilder.buffer_size",
                         ReversePathHierarchyTokenizerBuilder.DEFAULT_BUFFER_SIZE,
                         builder.bufferSize);
        assertExactValue("PathHierarchyTokenizerBuilder.delimiter",
                         ReversePathHierarchyTokenizerBuilder.DEFAULT_DELIMITER,
                         builder.delimiter);
        assertExactValue("PathHierarchyTokenizerBuilder.replacement",
                         ReversePathHierarchyTokenizerBuilder.DEFAULT_REPLACEMENT,
                         builder.replacement);
        assertExactValue("PathHierarchyTokenizerBuilder.skip",
                         ReversePathHierarchyTokenizerBuilder.DEFAULT_SKIP,
                         builder.skip);
    }

    @Test(expected = IOException.class)
    public void testReversePathHierarchyTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail(
                "{type: \"reverse_path_hierarchy\", buffer_size: 246, delimiter: \"/\", replacent: \"%\", skip: 3}");
    }

    @Test
    public void testStandardTokenizerValidJSON() {
        String json = "{type: \"standard\", max_token_length: 246}";
        StandardTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                     StandardTokenizerBuilder.class,
                                                                     StandardTokenizer.class);
        assertExactValue("StandardTokenizerBuilder.maxTokenLength", 246, builder.maxTokenLength);
    }

    @Test
    public void testStandardTokenizerDefaultValues() {
        StandardTokenizerBuilder builder = assertBuilderAndTokenizer("{type: \"standard\"}",
                                                                     StandardTokenizerBuilder.class,
                                                                     StandardTokenizer.class);
        assertExactValue("ClassicTokenizerBuilder.maxTokenLength",
                         StandardTokenizerBuilder.DEFAULT_MAX_TOKEN_LENGTH,
                         builder.maxTokenLength);
    }

    @Test(expected = IOException.class)
    public void testStandardTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"standard\", max_token_ngth: 246}");
    }

    @Test
    public void testUAX29URLEmailTokenizerValidJSON() {
        String json = "{type: \"uax29_url_email\", max_token_length: 249}";
        UAX29URLEmailTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                          UAX29URLEmailTokenizerBuilder.class,
                                                                          UAX29URLEmailTokenizer.class);
        assertExactValue("UAX29URLEmailTokenizerBuilder.maxTokenLength", 249, builder.maxTokenLength);
    }

    @Test
    public void testUAX29URLEmailTokenizerDefaultValues() {
        String json = "{type: \"uax29_url_email\"}";
        UAX29URLEmailTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                          UAX29URLEmailTokenizerBuilder.class,
                                                                          UAX29URLEmailTokenizer.class);
        assertExactValue("UAX29URLEmailTokenizerBuilder.maxTokenLength",
                         UAX29URLEmailTokenizerBuilder.DEFAULT_MAX_TOKEN_LENGTH,
                         builder.maxTokenLength);
    }

    @Test(expected = IOException.class)
    public void testUAX29URLEmailTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"uax29_url_email\", max_token_lgth: 249}");
    }

    @Test
    public void testUnicodeWhitespaceTokenizerValidJSON() {
        String json = "{type:\"unicode_whitespace\"}";
        assertBuilderAndTokenizer(json, UnicodeWhitespaceTokenizerBuilder.class, UnicodeWhitespaceTokenizer.class);
    }

    @Test
    public void testWhitespaceTokenizerValidJSON() {
        String json = "{type:\"whitespace\"}";
        assertBuilderAndTokenizer(json, WhitespaceTokenizerBuilder.class, WhitespaceTokenizer.class);
    }

    @Test
    public void testWikipediaTokenizerValidJSON() {
        String json = "{type: \"wikipedia\", token_output: \"TOKENS_ONLY\", untokenized_types : [\"aaa\",\"bbb\"]}";
        WikipediaTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                      WikipediaTokenizerBuilder.class,
                                                                      WikipediaTokenizer.class);
        assertExactValue("WikipediaTokenizerBuilder.token_output",
                         WikipediaTokenizerBuilder.TokenOutputValue.TOKENS_ONLY,
                         builder.tokenOutput);
        assertExactValue("WikipediaTokenizerBuilder.untokenized_types",
                         Sets.newHashSet("aaa", "bbb"),
                         builder.untokenizedTypes);
    }

    @Test
    public void testWikipediaTokenizerDefaultValues() {
        String json = "{type: \"wikipedia\"}";
        WikipediaTokenizerBuilder builder = assertBuilderAndTokenizer(json,
                                                                      WikipediaTokenizerBuilder.class,
                                                                      WikipediaTokenizer.class);
        assertExactValue("WikipediaTokenizerBuilder.token_output",
                         WikipediaTokenizerBuilder.TokenOutputValue.TOKENS_ONLY,
                         builder.tokenOutput);
        assertExactValue("WikipediaTokenizerBuilder.untokenized_types", Sets.newHashSet(), builder.untokenizedTypes);
    }

    @Test(expected = IOException.class)
    public void testWikipediaTokenizerInvalidJSON() throws IOException {
        assertJsonParseFail("{type: \"wikipedia\", token_output: \"TOKENS_ONLY\", untoknized_types : [\"aaa\",\"bbb\"]}");
    }

    @Test(expected = IOException.class)
    public void testInvalidTokenizerType() throws IOException {
        assertJsonParseFail("{type: \"invalid_type\"}");
    }

}

File: plugin/src/test/scala/com/stratio/cassandra/lucene/schema/analysis/tokenizer/TokenizerBuilderTest.scala
/*
 * Copyright (C) 2014 Stratio (http://stratio.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.stratio.cassandra.lucene.schema.analysis.tokenizer

import com.stratio.cassandra.lucene.BaseScalaTest
import com.stratio.cassandra.lucene.common.JsonSerializer
import org.apache.lucene.analysis.core._
import org.apache.lucene.analysis.ngram.{NGramTokenizer, EdgeNGramTokenizer}
import org.apache.lucene.analysis.path.{ReversePathHierarchyTokenizer, PathHierarchyTokenizer}
import org.apache.lucene.analysis.pattern.PatternTokenizer
import org.apache.lucene.analysis.standard.{UAX29URLEmailTokenizer, StandardTokenizer, ClassicTokenizer}
import org.apache.lucene.analysis.th.ThaiTokenizer
import org.apache.lucene.analysis.util.TokenizerFactory
import org.apache.lucene.analysis.wikipedia.WikipediaTokenizer
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner

import scala.util.Try

/** Tests for [[TokenizerBuilder]].
  *
  * @author Juan Pedro Gilaberte `jpgilaberte@stratio.com`
  */
@RunWith(classOf[JUnitRunner])
class TokenizerBuilderTest extends BaseScalaTest{

  test("available tokenizer"){
    assert(TokenizerFactory.availableTokenizers().size() == 13)
  }

  type T = TokenizerBuilder[TokenizerFactory]

  def assertBuild(tokenizerBuilder: T, factoryClass: String, tokenizerClass: String) = {
    val tokenizerFactory = tokenizerBuilder.build
    val tokenizer = tokenizerFactory.create()
    assert(tokenizerFactory.getClass.getSimpleName == factoryClass)
    assert(tokenizer.getClass.getSimpleName == tokenizerClass)
  }

  def assertBuildException (tokenizerBuilder: T) = assertThrows[RuntimeException](tokenizerBuilder.build.create())

  test("ClassicTokenizerBuilder") {
    val jsonTest1 = """{"type":"classic", "max_token_length": 25}"""
    val jsonTest2 = """{"type":"classic"}"""
    val jsonTest3 = """{"type":"classic", "max_token_length": -25}"""
    val factoryName = "ClassicTokenizerFactory"
    val tokenizerName = "ClassicTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[ClassicTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[ClassicTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[ClassicTokenizerBuilder]))
  }

  test("EdgeNGramTokenizerBuilder") {
    val jsonTest1 = """{"type":"edge_ngram", "max_gram_size": 25, "min_gram_size": 10}"""
    val jsonTest2 = """{"type":"edge_ngram"}"""
    val jsonTest3 = """{"type":"edge_ngram", "max_gram_size": 25, "min_gram_size": -10}"""
    val factoryName = "EdgeNGramTokenizerFactory"
    val tokenizerName = "EdgeNGramTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[EdgeNGramTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[EdgeNGramTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[EdgeNGramTokenizerBuilder]))
  }

  test("KeywordTokenizerBuilder") {
    val jsonTest1 = """{"type":"keyword"}"""
    val factoryName = "KeywordTokenizerFactory"
    val tokenizerName = "KeywordTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[KeywordTokenizerBuilder]), factoryName, tokenizerName)
  }

  test("LetterTokenizerBuilder") {
    val jsonTest1 = """{"type":"letter"}"""
    val factoryName = "LetterTokenizerFactory"
    val tokenizerName = "LetterTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[LetterTokenizerBuilder]), factoryName, tokenizerName)
  }

  test("LowerCaseTokenizerBuilder") {
    val jsonTest1 = """{"type":"lower_case"}"""
    val factoryName = "LowerCaseTokenizerFactory"
    val tokenizerName = "LowerCaseTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[LowerCaseTokenizerBuilder]), factoryName, tokenizerName)
  }

  test("NGramTokenizerBuilder") {
    val jsonTest1 = """{"type":"ngram", "max_gram_size": 25, "min_gram_size": 10}"""
    val jsonTest2 = """{"type":"ngram"}"""
    val jsonTest3 = """{"type":"ngram", "max_gram_size": 25, "min_gram_size": -10}"""
    val factoryName = "NGramTokenizerFactory"
    val tokenizerName = "NGramTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[NGramTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[NGramTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[NGramTokenizerBuilder]))
  }

  test("PathHierarchyTokenizerBuilder") {
    val jsonTest1 = """{"type":"path_hierarchy", "reverse": false, "delimiter": "/", "replace": "%", skip: 3}"""
    val jsonTest2 = """{"type":"path_hierarchy"}"""
    val jsonTest3 = """{"type":"path_hierarchy", "reverse": false, "delimiter": "/", "replace": "%", skip: -3}"""
    val factoryName = "PathHierarchyTokenizerFactory"
    val tokenizerName = "PathHierarchyTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[PathHierarchyTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[PathHierarchyTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[PathHierarchyTokenizerBuilder]))

    val jsonTest4 = """{"type":"path_hierarchy", "reverse": true, "delimiter": "/", "replace": "%", skip: 3}"""
    val jsonTest5 = """{"type":"path_hierarchy", "reverse": true, "delimiter": "/", "replace": "%", skip: -3}"""
    val tokenizerNameReverse = "ReversePathHierarchyTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest4, classOf[PathHierarchyTokenizerBuilder]), factoryName, tokenizerNameReverse)
    assertBuildException(JsonSerializer.fromString(jsonTest5, classOf[PathHierarchyTokenizerBuilder]))
  }

  test("PatternTokenizerBuilder") {
    val jsonTest1 = """{"type":"pattern", pattern: "[a-z]", group: 0}"""
    val jsonTest2 = """{"type":"pattern", pattern: "[a-z]", group: 0}"""
    val jsonTest3 = """{"type":"pattern"}"""
    val factoryName = "PatternTokenizerFactory"
    val tokenizerName = "PatternTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[PatternTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[PatternTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[PatternTokenizerBuilder]))
  }

  test("StandardTokenizerBuilder") {
    val jsonTest1 = """{"type":"standard", "max_token_length": 25}"""
    val jsonTest2 = """{"type":"standard"}"""
    val jsonTest3 = """{"type":"standard", "max_token_length": -25}"""
    val factoryName = "StandardTokenizerFactory"
    val tokenizerName = "StandardTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[StandardTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[StandardTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[StandardTokenizerBuilder]))
  }

  test("ThaiTokenizerBuilder") {
    val jsonTest1 = """{"type":"thai"}"""
    val factoryName = "ThaiTokenizerFactory"
    val tokenizerName = "ThaiTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[ThaiTokenizerBuilder]), factoryName, tokenizerName)
  }

  test("UAX29URLEmailTokenizerBuilder") {
    val jsonTest1 = """{"type":"uax29_url_email", "max_token_length": 25}"""
    val jsonTest2 = """{"type":"uax29_url_email"}"""
    val jsonTest3 = """{"type":"uax29_url_email", "max_token_length": -25}"""
    val factoryName = "UAX29URLEmailTokenizerFactory"
    val tokenizerName = "UAX29URLEmailTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[UAX29URLEmailTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[UAX29URLEmailTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[UAX29URLEmailTokenizerBuilder]))
  }

  test("WhiteSpaceTokenizerBuilder") {
    val jsonTest1 = """{"type":"whitespace", "rule": "java"}"""
    val jsonTest2 = """{"type":"whitespace"}"""
    val jsonTest3 = """{"type":"whitespace", "rule": "failure"}"""
    val factoryName = "WhitespaceTokenizerFactory"
    val tokenizerName = "WhitespaceTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[WhitespaceTokenizerBuilder]), factoryName, tokenizerName)
    assertBuild(JsonSerializer.fromString(jsonTest2, classOf[WhitespaceTokenizerBuilder]), factoryName, tokenizerName)
    assertBuildException(JsonSerializer.fromString(jsonTest3, classOf[WhitespaceTokenizerBuilder]))

    val jsonTest4 = """{"type":"whitespace", "rule": "unicode"}"""
    val tokenizerNameUnicode = "UnicodeWhitespaceTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest4, classOf[WhitespaceTokenizerBuilder]), factoryName, tokenizerNameUnicode)
  }

  test("WikipediaTokenizerBuilder") {
    val jsonTest1 = """{"type":"wikipedia"}"""
    val factoryName = "WikipediaTokenizerFactory"
    val tokenizerName = "WikipediaTokenizer"
    assertBuild(JsonSerializer.fromString(jsonTest1, classOf[WikipediaTokenizerBuilder]), factoryName, tokenizerName)
  }
}
